ci/ci/envoy.py

```diff
@@ -101,22 +115,35 @@ def gateway_default_host(service: str) -> dict:
     }
 
 
-def gateway_internal_host(services_per_namespace: Dict[str, List[str]]) -> dict:
+def gateway_internal_host(
+    internal_namespaces: List[NamespaceRoutingConfig],
+) -> dict:
+    routes = [
+        {
+            'match': {'path_separated_prefix': f'/{namespace.name}/{service}'},
+            'route': route_to_cluster(f'{namespace.name}-{service}'),
+            'typed_per_filter_config': {
+                'envoy.filters.http.local_ratelimit': rate_limit_config(service),
+            },
+        }
+        for namespace in internal_namespaces
+        for service in namespace.deployed_services
+    ]
+
+    oauth_callback_rewrites = [
+        {
+            'match': {'prefix': urlparse(namespace.oauth2_callback_url).path},
+            'route': {'cluster': f'{namespace.name}-auth', 'prefix_rewrite': f'/{namespace.name}/auth/oauth2callback'},
+        }
+        for namespace in internal_namespaces
+    ]
+    routes.extend(oauth_callback_rewrites)
```

Comment:
Do I understand correctly, that this is only necessary because oauth2 callback URLs are named by greek letters (due to the programmatic OAuth2 create/destroy limitations in GCP) and we need to direct the greek letter to the associated namespace (which has a developer-chosen name)?